### PR TITLE
Don't allow actions whose parameter types don't match the call

### PR DESCRIPTION
**Pull request by FeepingCreature**
ping @belka-ew @linkrope

### CODE REVIEW

File: source/dmocks/action.d [hunk 1]
```diff
@@ -72,7 +62,10 @@ struct Actor
             }
             else
             {
-                return Rope(No.pass, ActionStatus.FailBadAction);
+                import std.format : format;
+
+                throw new Error(format!"cannot call action: type %s does not match argument type %s"
```
Comment (belka-ew) [inline source/dmocks/action.d]: muss es nicht eine Exception (und vielleicht sogar eine benutzerdefinierte)?

Comment (FeepingCreature) [reply source/dmocks/action.d]: Ne, weil es ein harter Code-Fehler ist, von dem man sich nicht erholen können sollte.

File: source/dmocks/mocks.d [hunk 2]
```diff
@@ -772,6 +772,19 @@ unittest
     assert (calledPayload);
 }
 
+@("delegate payload with mismatching parameters")
+unittest
+{
+    Mocker r = new Mocker();
```
Comment (linkrope) [inline source/dmocks/mocks.d]: (at least) three styles are mixed in the unit-test functions:
- the mocker is called `m`
- the mocker is called `r` (maybe due to a conflict with the mocked something that is called `m`)
- the mocker is called `mocker`

at least for new unit tests, I would prefer

    auto mocker = new Mocker();

File: source/dmocks/mocks.d [hunk 3]
```diff
@@ -772,6 +772,19 @@ unittest
...
+    auto o = r.mock!(SimpleObject);
+
+    //o.print;
+    r.expect(o.print).action((int) { });
+    r.replay();
+
+    assertThrown!Error(o.print);
```
Comment (linkrope) [inline source/dmocks/mocks.d]: We already use `@("...")` from unit-threaded, but we don't use unit-threaded for the unit test.
So, we cannot use, for example, the better `shouldThrowWithMessage`.

**Review by linkrope — APPROVED** [review]

**Review by belka-ew — APPROVED** [review]

**Pull request merged**